src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+│                      │
+└──────────────────────┘
+```
+
+We propose that such a reduction could be done using the LLVM passes infrastructure to compose a profile
+which would map the QIR to a subset of the available instructions with any required constraints.
+
+## Feasibility Study
+
+In order to demonstrate feasibility of this proposal, we have built a proof-of-concept prototype based on LLVM passes which allows transformation from a generic QIR into one which does not have support for dynamic qubit allocation.
+
+This transformation is considered to be the smallest, non-trivial case of QIR transformation we can perform which demonstrates the feasibility of this proposal.
+
+To demonstrate the feasibility of this proposal, we use Q# as a frontend and will attempt to map the following code
+
+```
+namespace Feasibility
+{
+    open Microsoft.Quantum.Intrinsic;
+
+    @EntryPoint()
+    operation QubitMapping() : Unit {
+        use qs = Qubit[3];
+        for q in 8..10 {
+            X(qs[q - 8]);
+        }
+    }
+}
+```
+
+to the [base profile](https://github.com/microsoft/qsharp-language/blob/ageller/profile/Specifications/QIR/Base-Profile.md). We will do so using a combination of existing LLVM passes and custom written passes which are specific to the QIR. The above code is interesting as it is not base profile compliant with regards to two aspects: 1) Qubit allocation is not allowed and 2) arithmetic operations are not supported. Using the Q# QIR generator, the `QubitMapping` functions body becomes:
+
+```
+define internal void @Feasibility__QubitMapping__body() {
+entry:
+  %qs = call %Array* @__quantum__rt__qubit_allocate_array(i64 3)
+  call void @__quantum__rt__array_update_alias_count(%Array* %qs, i32 1)
+  br label %header__1
+
+header__1:                                        ; preds = %exiting__1, %entry
+  %q = phi i64 [ 8, %entry ], [ %4, %exiting__1 ]
+  %0 = icmp sle i64 %q, 10
+  br i1 %0, label %body__1, label %exit__1
+
+body__1:                                          ; preds = %header__1
+  %1 = sub i64 %q, 8
+  %2 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %qs, i64 %1)
+  %3 = bitcast i8* %2 to %Qubit**
+  %qubit = load %Qubit*, %Qubit** %3, align 8
+  call void @__quantum__qis__x__body(%Qubit* %qubit)
+  br label %exiting__1
+
+exiting__1:                                       ; preds = %body__1
+  %4 = add i64 %q, 1
+  br label %header__1
+
+exit__1:                                          ; preds = %header__1
+  call void @__quantum__rt__array_update_alias_count(%Array* %qs, i32 -1)
+  call void @__quantum__rt__qubit_release_array(%Array* %qs)
+  ret void
+}
+```
+
+After applying the our [demo profile transformation](https://github.com/troelsfr/qsharp-compiler/tree/feature/profile-pass/src/Passes/examples/QubitAllocationAnalysis), the QIR is reduced to:
+
+```
+define void @Feasibility__QubitMapping__Interop() local_unnamed_addr #0 {
+entry:
+  call void @__quantum__qis__x__body(%Qubit* null)
```

Comment:
Hmm, it is a bit suboptimal that qubit index 0 becomes a null pointer. What happens if the pointer is cast back to int64 again? Does it give you 0L?

Reply:
Yes. `null` is LLVMs doing and I do not think there is much to do about it, but I have not looked.